File: javascript/packages/linter/src/cli/fixability.ts
import type { LintOffense, RuleClass } from "../types.js"

export interface Fixability {
  autocorrectable: boolean
  unsafeAutocorrectable: boolean
}

const NOT_FIXABLE: Fixability = { autocorrectable: false, unsafeAutocorrectable: false }

export function fixabilityFor(offense: LintOffense, ruleClass: RuleClass | undefined): Fixability {
  if (!ruleClass) return NOT_FIXABLE
  if (ruleClass.autofixRequiresContext === true && !offense.autofixContext) return NOT_FIXABLE

  const correctable = ruleClass.autocorrectable === true
  const unsafe = ruleClass.unsafeAutocorrectable === true || offense.autofixContext?.unsafe === true

  if (!correctable && !unsafe) return NOT_FIXABLE

  return { autocorrectable: correctable && !unsafe, unsafeAutocorrectable: unsafe }
}

File: rust/herb-linter/src/fixability.rs
use crate::offense::Offense;
use crate::rule::{AnyRule, Rule};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct Fixability {
  pub autocorrectable: bool,
  pub unsafe_autocorrectable: bool,
}

/// A rule can declare every one of its offenses unsafe to fix, and an offense
/// can declare itself unsafe even when the rule is generally safe, so both are
/// consulted before an offense counts as safely correctable.
pub fn fixability_for(offense: &Offense, rule: Option<&AnyRule>) -> Fixability {
  let rule = match rule {
    Some(rule) => rule,
    None => return Fixability::default(),
  };

  let correctable = rule.autocorrectable();
  let unsafe_correctable = rule.unsafe_autocorrectable() || offense.unsafe_fix;

  if !correctable && !unsafe_correctable {
    return Fixability::default();
  }

  Fixability {
    autocorrectable: correctable && !unsafe_correctable,
    unsafe_autocorrectable: unsafe_correctable,
  }
}
